### PR TITLE
aws_http_headers

### PR DESCRIPTION
**Pull request by graebm**
Nice headers API to match the one we came up with in the bindings.
Ref: [awscrt.http.HttpHeaders](https://github.com/awslabs/aws-crt-python/blob/ef8e18c559f25cf26f196c091d7f47827dc2cf43/awscrt/http.py#L215)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/http/request_response.h [hunk 1]
```diff
@@ -272,6 +280,113 @@ struct aws_http_request_handler_options {
...
+ *
+ * - Headers with different names could be in any order, relative to one another.
+ *   If "A: one" is seen before "B: bee" in one iteration, you might see "B: bee" before "A: one" on the next.
+ *
+ * AWS_ERROR_INVALID_INDEX is raised if the index is invalid.
+ */
+AWS_HTTP_API
+int aws_http_headers_get_index(
+    const struct aws_http_headers *headers,
+    struct aws_http_header *out_header,
```
Comment (ColdenCullen) [inline include/aws/http/request_response.h]: plz reverse

File: include/aws/http/request_response.h [hunk 2]
```diff
@@ -272,6 +280,113 @@ struct aws_http_request_handler_options {
...
+    size_t index);
+
+/**
+ * Get the first value for this name, ignoring any additional values.
+ * AWS_ERROR_HTTP_HEADER_NOT_FOUND is raised if the name is not found.
+ */
+AWS_HTTP_API
+int aws_http_headers_get(
+    const struct aws_http_headers *headers,
+    struct aws_byte_cursor *out_value,
```
Comment (ColdenCullen) [inline include/aws/http/request_response.h]: just yeet it

File: source/request_response.c [hunk 3]
```diff
@@ -30,9 +31,279 @@ enum {
...
+    if (name.len == 0) {
+        return aws_raise_error(AWS_ERROR_HTTP_INVALID_HEADER_NAME);
+    }
+
+    size_t total_len;
+    if (aws_add_size_checked(name.len, value.len, &total_len)) {
+        return AWS_OP_ERR;
+    }
+
+    struct aws_http_header header = {name, value};
```
Comment (ColdenCullen) [inline source/request_response.c]: holy cow I forgot C had this syntax.

File: source/request_response.c [hunk 4]
```diff
@@ -30,9 +31,279 @@ enum {
...
+        struct aws_http_header *header = NULL;
+        aws_array_list_get_at_ptr(&headers->array_list, (void **)&header, i);
+        if (aws_http_header_name_eq(header->name, name)) {
+            s_http_headers_erase_index(headers, i);
+            --end_index;
+            erased_any = true;
+        } else {
+            ++i;
+        }
+    }
```
Comment (ColdenCullen) [inline source/request_response.c]: make go backwards

File: include/aws/http/request_response.h (hunk 1 above)
Comment (graebm) [reply include/aws/http/request_response.h]: Originally I had it like that, with out-params last, then changed to be like the aws_array_list API.
Seems like we all like it better the other way, switching

File: source/request_response.c [hunk 5]
```diff
@@ -180,17 +180,18 @@ int aws_http_headers_erase_index(struct aws_http_headers *headers, size_t index)
 /* Erase entries with name, stop at end_index */
 static int s_http_headers_erase(struct aws_http_headers *headers, struct aws_byte_cursor name, size_t end_index) {
     bool erased_any = false;
-    for (size_t i = 0; i < end_index;) {
-        struct aws_http_header *header = NULL;
+    struct aws_http_header *header = NULL;
+
+    /* Iterating in reverse is simpler */
+    for (size_t n = end_index; n > 0; --n) {
+        size_t i = n - 1;
```
Comment (ColdenCullen) [inline source/request_response.c]: CONST

File: include/aws/http/request_response.h [hunk 6]
```diff
@@ -293,11 +301,26 @@ AWS_EXTERN_C_BEGIN
...
 AWS_HTTP_API
 struct aws_http_headers *aws_http_headers_new(struct aws_allocator *allocator);
 
+/**
+ * Acquire a hold on the object, preventing it from being deleted until
+ * aws_http_headers_release() is called by all those with a hold on it.
+ */
 AWS_HTTP_API
-void aws_http_headers_destroy(struct aws_http_headers *headers);
+void aws_http_headers_acquire_hold(struct aws_http_headers *headers);
```
Comment (bretambrose) [inline include/aws/http/request_response.h]: Why not just acquire?

Comment (ColdenCullen) [reply include/aws/http/request_response.h]: Yeah, Mike? WHY NOT JUST ACQUIRE.

File: source/request_response.c (hunk 5 above)
Comment (graebm) [reply source/request_response.c]: OK

File: include/aws/http/request_response.h (hunk 6 above)
Comment (graebm) [reply include/aws/http/request_response.h]: aws_channel has `acquire_hold()` and `release_hold()`, but I just searched and see that vanilla `acquire()` and `release()` are more prevalent

**Pull request merged**